Android updates

magneticflux- opened the pull request:
Closes #5. Travis CI still needs to be enabled on the OpenFTC Organization and on this particular repository.

---
A lot of work needs to be done for #15.
#4 should be only a few lines now that the build has been updated to Android Gradle plugin 3.0.0.

Request changes if there are any issues!

magneticflux- committed:
Updated Gradle to 4.2.1, Android Gradle plugin to 3.0.0, compile SDK to 26

magneticflux- committed:
Added Java 8 support to TeamCode

magneticflux- committed:
Added .travis.yml

magneticflux- committed:
Added Travis CI badge

magneticflux- committed:
Replaced compile with implementation configuration (except RobotCore, which leaks so many dependencies)

magneticflux- committed:
Fixed all critical lint errors

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java:
```diff
@@ -33,11 +33,10 @@
  * @author lizlooney@google.com (Liz Looney)
  */
 public abstract class AbstractProgrammingModeActivity extends ThemedActivity {
+    private final Timer timer = new Timer();
```
Spacing issue

qwertychouskie replied:
Also why does this need to be moved?

magneticflux- replied:
I accidentally ran my formatter on changed files, I'll undo the spacing and order changes.

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java:
```diff
@@ -179,7 +178,7 @@ public void run() {
    */
   private void setTypeface() {
-    Typeface typeface = Typeface.create(Typeface.SERIF, 0);
+      Typeface typeface = Typeface.create(Typeface.SERIF, Typeface.NORMAL);
```
Spacing again

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -44,12 +44,10 @@
   private static final AtomicReference<String> nameOfOpModeLoadedIntoWebView = new AtomicReference<String>();
   private static final Map<String, Access> javascriptInterfaces = new ConcurrentHashMap<String, Access>();
-
+    private static Activity activity;
```
Here too

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -67,6 +65,61 @@
     logPrefix = LOG_PREFIX + "\"" + project + "\" - ";
   }
 
+    /**
```
And this whole section

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -282,16 +335,17 @@ public void run() {
-  private void addJavascriptInterfaces(HardwareItemMap hardwareItemMap) {
-    addJavascriptInterfacesForIdentifiers();
-    addJavascriptInterfacesForHardware(hardwareItemMap);
+    @SuppressLint("JavascriptInterface")
```
And here

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -282,16 +335,17 @@ public void run() {
+            String identifier = entry.getKey();
+            Access access = entry.getValue();
+            webView.addJavascriptInterface(access, identifier);
+        }
```
And here

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
-      addOpModeRegistrar();
-    }
+    private class BlocksOpModeAccess extends Access {
+        private final Object scriptFinishedLock;
```
And here

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
+        private BlocksOpModeAccess(String identifier, Object scriptFinishedLock, AtomicBoolean scriptFinished) {
+            super(BlocksOpMode.this, identifier);
+            this.scriptFinishedLock = scriptFinishedLock;
+            this.scriptFinished = scriptFinished;
```
And here

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
+        @SuppressWarnings("unused")
+        @JavascriptInterface
+        public void scriptStarting() {
+            RobotLog.i(getLogPrefix() + "scriptStarting");
```
And here

qwertychouskie commented on Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java:
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
+                scriptFinished.set(true);
+                scriptFinishedLock.notifyAll();
+            }
+    }
```
Here too

qwertychouskie commented on FtcCommon/src/main/res/layout/lynx_module_configure_address_spin_item.xml:
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
-<CheckedTextView android:id="@android:id/text1"
+<CheckedTextView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@android:id/text1"
```
Also wrong spacing

qwertychouskie commented on RobotCore/build.gradle:
```diff
@@ -18,9 +16,8 @@ android {
 
 dependencies {
-    compile fileTree(include: ['*.jar'], dir: 'libs') //mark as api once we update the Android plugin to 3.0+
+    api fileTree(include: ['*.jar'], dir: 'libs') //mark as api once we update the Android plugin to 3.0+
```
You can probably remove the comment here now

qwertychouskie commented on build.gradle:
```diff
@@ -5,10 +5,11 @@
  */
 buildscript {
     repositories {
+        google()
```
Is this extra repo really needed?

magneticflux- replied:
It's required for Android Gradle plugin 3.0.0

qwertychouskie reviewed (changes requested):
A few nitpicks as follows:

FROGbots-4634 commented:
Changing the targetSdk version to 26 will cause all kinds of crashes because >=23 will enable runtime permissions

magneticflux- commented:
@FROGbots-4634 Per the [documentation](https://developer.android.com/training/permissions/requesting.html):
>If the device is running Android 5.1 or lower, or your app's target SDK is 22 or lower: If you list a dangerous permission in your manifest, the user has to grant the permission when they install the app; if they do not grant the permission, the system does not install the app at all.

>If the device is running Android 6.0 or higher, and your app's target SDK is 23 or higher: The app has to list the permissions in the manifest, and it must request each dangerous permission it needs while the app is running. The user can grant or deny each permission, and the app can continue to run with limited capabilities even if the user denies a permission request.

Are there teams that are running on a phone with API >= 23? The phones my team had were all API 19 or 21. I believe the official FTC documentation says that phones above Android Lollipop (API 21) aren't supported.

magneticflux- committed:
Reverted formatting changes

magneticflux- committed:
Updated RobotCore api comment

magneticflux- committed:
Corrected repository ordering

qwertychouskie reviewed (approved):
Looking good!

FROGbots-4634 commented:
@magneticflux- 
1. Yes, there are teams running >= API 23. In fact, the phone officially recommend by FIRST for this year is running Marshmallow.

2. Android Lollipop is not supported due to a USB OTG bug that only allows it to communicate with one downstream USB device at a time. This was fixed in Marshmallow.

3. Regarding >= 21 not being supported, that's quite incorrect. In fact, per the game manual for this year, the ZTE Speed is the only one of the allowed list of phones that is not *required* to be on Marshmallow or higher.

magneticflux- commented:
@FROGbots-4634 What about requesting Camera permissions on app launch? That would prevent any Vuforia errors from being denied the camera. Are there any other permission errors?

FROGbots-4634 commented:
@magneticflux- You'd also need to request external storage permission.

Requesting permissions at all would add a lot of completely to MainActivity, something we do not want to do. The goal of OpenFTC is to add features as unobtrusively as possible, modifying as little of FIRST's code as possible. Any significant changes to MainActivity will make any future merges from FIRST's upstream very complicated and just a pain in general.

magneticflux- commented:
@FROGbots-4634 I added a permission request for Camera and Read/Write External Storage in about 5 lines of code. It doesn't give any user-friendly info, but it does allow permissions.

Additionally, the app was compiled with API 23 before I updated it to API 26. That means that the permission changes were in place before I increased the API further.

FROGbots-4634 commented:
@magneticflux- incorrect, compileSdk is different from targetSdk.

You can have a compileSdk of 26 and still target 19.

magneticflux- committed:
Changed targetSdkVersion to 19 for all projects

magneticflux- commented:
@FROGbots-4634 Thanks, I missed that detail!

The next issue to be tackled is the `.jar` dependencies of `RobotCore`, namely the `javac.jar` dependency. @NoahAndrews What was your reasoning behind including the entire Java compiler?

FROGbots-4634 commented:
@magneticflux- He didn't include it - FIRST did. It's needed to compile user code written with OnBotJava

noahAndrews committed:
Merge branch 'master' into android-updates

ItsHarper reviewed (changes requested):
Sorry this took me so long to get to. 

I'll merge this, but I'd like you to add documentation to the readme that this will require Android Studio 3.0 and updated SDK components. I'd really like to see screenshots included of what the errors will look like if you try to build this in Android Studio 2.3.x or with an Android SDK that doesn't have platform 26 installed.

magneticflux- committed:
Added Android Studio 3.0 note in readme

magneticflux- committed:
Merge branch 'master' into android-updates

NoahAndrews commented:
Thank you for submitting this, but I'm going to close it now. We ended up moving to gradle plugin 3 in a different commit, and the Travis stuff can wait for now.
